README.md
# superagent-toolkit

> [Superagent][superagent] plugin, providing frequently used tools, like reporting message of current request, directly retrieving specific property of response, etc.

[![Build Status](https://travis-ci.org/lakca/superagent-toolkit.svg?branch=master)](https://travis-ci.org/lakca/superagent-toolkit)
[![codecov](https://codecov.io/gh/lakca/superagent-toolkit/branch/master/graph/badge.svg)](https://codecov.io/gh/lakca/superagent-toolkit)

## Usage

### Agent
```js
const superagent = require('superagent')
const toolkit = require('@lakca/superagent-toolkit')

const agent = new superagent.Agent()

agent.use(toolkit(options))

agent
  .get('/test')
  .configure('...', '...') // change config
```

### General
```js
const superagent = require('superagent')
const toolkit = require('@lakca/superagent-toolkit')

superagent
  .get('/test')
  .use(toolkit(options))
```

## Options

### `options.configure` {function}

Set configure function, which can be called to change `config` later.

### `options.report` {(error: Error, message: string, response: superagent.Response): void}

Function to receive message of request.

**Note**: it will benefit from using this function for global notification of request status.

### `options.clientError` {true|string|string[]|function}

Dealing with `4xx` except `401` http code. If value is:

- `true`, report message retrieved using `config.messageProperty`;
- `string`, report message using the string instead;
- `string[]`, report message using ths first not nullable(`!= null`) property retrieving via the array element one by one, such as `['body.message', 'header.statusMessage']`;
- `function`, report message using returned value if which is not void.

### `options.serverError` {true|string|string[]|function}

Dealing with `5xx` http code. Similar as `config.clientError`.

### `options.unauthorized` {true|string|string[]|function}

Dealing with `401` http code. If value is:

- `true`, reload current window;
- `string`, redirect using the string;
- `string[]`, redirect using ths first not nullable(`!= null`) property retrieving via the array element one by one, such as `['body.location']`;
- `function`, redirect using returned value if which is not void, **so you can provide your own redirection in the function without returning value.**

### `options.messageProperty` {string=res.statusMessage}

Report message property.

when http code is `2xx`, using this to retrieve message.

also, this is the default message property for `true` value of `config.clientError` and `config.serverError`.

**Note**: support descendant property, such as `body.message`

### `options.returnProperty` {string}

Response property expected to be returned instead of the whole response object.

**Note**: support descendant property, such as `body.data`

### `options.swallowError` {boolean=false}

Silence error, no request/response error will be thrown since [superagent][superagent] does that by default.

**Note**: this is useful when we handle request status by general handler `config.report`, then focusing on handling business data in dealing with your business logic.

### `options.debug` {boolean=false}

log to console.

### `options.prefix` {string}

### `options.suffix` {string}

### `options.jsonp` {object}

Read more in [superagent-jsonp][jsonp]

### `options.jsonp.callbackParam` {string=callback}

### `options.jsonp.callbackName` {string}

### `options.jsonp.timeout` {number=1000}

## LICENSE

MIT

[superagent]: https://github.com/visionmedia/superagent
[jsonp]: https://github.com/lamp/superagent-jsonp


History.md
# 1.0.0

- support `configure`
- support `report`
- support `messageProperty`
- support `returnProperty`
- support `clientError`
- support `serverError`
- support `unauthorized`
- support `swallowError`
- support `prefix`
- support `suffix`
- support `jsonp`
- support `debug`


test.js
const assert = require('assert')
const sp = require('superagent')
const tool = require('./index')

const userAgent = '5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.81 Safari/537.36'

const followers = sp.agent()
  .set('user-agent', userAgent)
  .use(tool({
    debug: true,
    prefix: 'http://api.github.com/users/',
    suffix: '/followers',
    clientError: true,
    serverError: true,
    messageProperty: 'status',
    unauthorized: true,
    swallowError: false,
    returnProperty: 'body',
    configure: 'config',
    jsonp: null,
    report(err, msg, res) {
      if (res) {
        assert.ok(msg === undefined || typeof msg === 'number', 'report message should be a number or undefined, since `messageProperty` is status')
      } else {
        assert.ok(typeof msg === 'string')
      }
      assert.ok(err === null || err instanceof Error)
      assert.ok(res === null || res instanceof sp.Response)
    }
  }))

followers.get('lakca').config('swallowError', true).end(function(err, res) {
  assert.ok(err === null, 'err should be always null, since `swallowError` is true.')
  assert.ok(!res || Array.isArray(res), 'res should be an array, since `returnProperty` is body.')
})


test/main.js
const { origin } = require('./support/server')
const superagent = require('superagent')
const toolkit = require('../index')
const should = require('should')

function mockWindow() {
  global.window = {
    location: {
      _reloaded: false,
      href: '',
      reload() {
        this._reloaded = true
      }
    }
  }
}

function mockConsole() {
  global.console = {
    log(v) {
      this._log = v
    },
    error(v) {
      this._error = v
    }
  }
}

describe('main', function() {

  describe('default', function() {
    const agent = superagent.agent().use(toolkit())

    it('default: ok', done => {
      agent.get(origin + '/status/200')
        .end((err, res) => {
          should(err).null()
          should(res.ok).true()
          done()
        })
    })

    it('default: client error', done => {
      agent.get(origin + '/status/400')
        .end((err, res) => {
          should(err).instanceof(Error)
          should(res.clientError).true()
          done()
        })
    })

    it('default: server error', done => {
      agent.get(origin + '/status/500')
        .end((err, res) => {
          should(err).instanceof(Error)
          should(res.serverError).true()
          done()
        })
    })

    it('default: destroy connection', done => {
      agent.get(origin + '/status/close')
        .end((err, res) => {
          err.should.instanceof(Error)
          should(res).undefined()
          done()
        })
    })
  })

  describe('opt', () => {
    const agent = superagent.agent().use(toolkit({
      configure: 'cfg',
      clientError: ['body.message'],
      serverError: 'Sorry, please try later!',
      unauthorized: '/login',
      returnProperty: 'body.data',
      swallowError: true,
      prefix: origin,
      suffix: '/' + escape('this is a message.'),
      debug: false,
      report(err, msg, res) {
        should(msg).eql(res.res.statusMessage)
      }
    }))

    it('2xx', done => {
      agent.get('/status/200')
        .end((err, res) => {
          should(err).be.null()
          should(res).eql({ greet: 'Hello World!' })
          done()
        })
    })

    it('3xx', done => {
      agent.get('/status/301')
        .redirects(0)
        .cfg('returnProperty', false)
        .end((err, res) => {
          should(err).be.null()
          should(res.headers.location).eql('/redirect')
          done()
        })
    })

    it('clientError (4xx)', done => {
      agent.get('/status/400')
        .cfg('report', (err, msg, res) => {
          should(msg).eql(res.body.message)
          should(res.clientError).be.true()
        })
        .end((err, res) => {
          should(err).be.null()
          done()
        })
    })

    it('serverError (5xx)', done => {
      agent.get('/status/500')
        .cfg('serverError', 'Sorry!')
        .cfg('report', (err, msg, res) => {
          should(msg).eql('Sorry!')
          should(res.serverError).be.true()
        })
        .end((err, res) => {
          should(err).be.null()
          done()
        })
    })

    it('serverError (5xx)', done => {
      agent.get('/status/503')
        .cfg('serverError', res => 'Oh, Sorry!')
        .cfg('report', (err, msg, res) => {
          should(msg).eql('Oh, Sorry!')
          should(res.serverError).be.true()
        })
        .end((err, res) => {
          should(err).be.null()
          done()
        })
    })

    it('error without response', done => {
      agent.get('')
        .cfg('report', (err, msg) => should(msg).eql(err.message))
        .abort()
        .end((err, res) => {
          should(err).be.null()
          should(res).be.undefined()
          done()
        })
    })

    it('jsonp', done => {
      mockWindow()
      const r = agent.get('/status/200')
        .cfg('jsonp', { callbackName: 'expectedCallbackName' })
      should(() => r.end()).be.throw(/document/)
      should(r._jsonp).property('callbackName', 'expectedCallbackName')
      done()
    })

    describe('unauthorized', () => {

      beforeEach(mockWindow)

      it('string', done => {
        agent.get('/status/401')
          .end((err, res) => {
            should(err).be.null()
            should(window.location.href).eql('/login')
            done()
          })
      })
      it('true', done => {
        agent.get('/status/401')
          .cfg('unauthorized', true)
          .end((err, res) => {
            should(err).be.null()
            should(window.location._reloaded).be.true()
            done()
          })
      })
      it('properties', done => {
        agent.get('/status/401')
          .cfg('unauthorized', ['a.b', 'c.d.e'])
          .end((err, res) => {
            should(err).be.null()
            should(window.location.href).be.undefined()
            done()
          })
      })
      it('function', done => {
        agent.get('/status/401')
          .cfg('unauthorized', res => '/sign')
          .end((err, res) => {
            should(err).be.null()
            should(window.location.href).eql('/sign')
            done()
          })
      })
    })
  })

  it('conflict configure method', done => {
    should(() => superagent.get('/').use(toolkit({ configure: 'set' }))).be.throw()
    done()
  })

  after(() => {
    describe('debug', () => {
      mockConsole()
      it('debug', done => {
        superagent.get('/')
          .use(toolkit({ debug: true }))
          .end(() => done())
        should(console._log).property('debug', true)
      })
    })
  })
})


test/support/server.js
const app = module.exports = require('express')()
app.port = 9001
app.origin = 'http://localhost:9001'

app.get(['/status/:status', '/status/:status/:message'], (req, res) => {
  if (req.params.status === 'close') {
    return req.destroy()
  }
  if (req.params.status === '301') {
    return res.redirect('/redirect')
  }
  res.status(+req.params.status)
    .set('x-message', req.params.message)
    .send({
      message: req.params.message,
      data: {
        greet: 'Hello World!'
      }
    })
})

app.listen(app.port)


index.js
const jsonp = require('superagent-jsonp')

/**
 * @callback report
 * @param {Error|null} error
 * @param {any} message reporting message
 * @param {Response} response
 * @return {void}
 */
/**
 * @typedef {import('superagent').SuperAgentRequest} SuperAgentRequest
 * @typedef {import('superagent').Response} Response
 * @typedef {string[]} SubProps using ths first not nullable(`!= null`) property retrieving via the elements one by one, supporting descendant property, such as `body.message`.
 */

/**
 * @typedef config
 *
 *
 * @property {string} [configure=configure] Set configure function, which can be called to change `config` later.
 *
 *
 * @property {boolean|string|SubProps|((response: Response) => string)} [clientError] dealing with http code is `4xx`: report.
 *
 * if value is:
 * - `true`, report message retrieved using `config.messageProperty`;
 * - `string`, report message using the string instead;
 * - `string[]`, report message using ths first not nullable(`!= null`) property retrieving via the array element one by one, such as ['body.message', 'header.statusMessage'];
 * - `function`, report message using returned value if which is not void.
 *
 *
 * @property {boolean|string|SubProps|((response: Response) => string)} [serverError] dealing with http code is `5xx`: report.
 *
 * if value is:
 * - `true`, report message retrieved using `config.messageProperty`;
 * - `string`, report message using the string instead;
 * - `string[]`, report message using ths first not nullable(`!= null`) property retrieving via the array element one by one, such as ['body.message', 'header.statusMessage'];
 * - `function`, report message using returned value if which is not void.
 *
 *
 * @property {boolean|string|SubProps|((response: Response) => string)} [unauthorized] dealing with http code is `401`: redirection.
 *
 * If value is:
 * - `true`, reload current window;
 * - `string`, redirect using the string;
 * - `string[]`, redirect using ths first not nullable(`!= null`) property retrieving via the array element one by one, such as ['body.location'];
 * - `function`, redirect using returned value if which is not void.
 *
 *
 * @property {string} [messageProperty=res.statusMessage] property name of response state message.
 *
 * **Note**: support descendant property, such as `body.message`
 *
 *
 * @property {report} [report] report message.
 *
 * **Note**: it will benefit from using this function for global notification.
 *
 *
 * @property {string} [returnProperty] response property expected to be returned instead of the whole response object.
 *
 * **Note**: support descendant property, such as `body.data`
 *
 *
 * @property {boolean} [swallowError] silence error.
 * no request/response error will be thrown since `superagent` does that by default.
 *
 *
 * @property {object} [jsonp] jsonp request, read more in `superagent-jsonp`
 * @property {string} [jsonp.callbackParam=callback] callback name in request query.
 * @property {string} [jsonp.callbackName] callback name returned should be.
 * @property {number} [jsonp.timeout=1000] millisecond.
 *
 *
 * @property {boolean} [debug] log to console.
 *
 *
 * @property {string} [prefix] url prefix
 *
 *
 * @property {string} [suffix] url suffix
 */

/**
 * Get sub property of object.
 *
 * @private
 *
 * @author lakca<912910011@qq.com>
 * @param {boolean} debug enable log (console)
 * @param {{[prop: string]: *,[prop: number]: *}} obj object which property is retrieved from
 * @param {string} description a string describing property location, such as 'body.data[0]'
 * @param {any} [defaultValue] default value is returned, if property is undefined.
 * @returns {any}
 */
function getSubProperty(debug, obj, description, defaultValue) {
  let value
  try {
    value = eval(`obj.${description}`)
    /* istanbul ignore if */
    if (value === undefined) value = defaultValue
  } catch (e) {
    /* istanbul ignore if */
    if (debug) console.error(e)
    value = defaultValue
  }
  return value
}

/**
 * @template T
 * @template {(element: T) => any} K
 * @param {T[]} arr
 * @param {K} fn
 * @return {ReturnType<K>}
 */
function returnNonNullable(arr, fn) {
  let i = 0, r
  while (i < arr.length) {
    r = fn(arr[i++])
    if (r != null) return r
  }
  return r
}

/**
 * @param {config} [config]
 */
module.exports = function(config) {
  config = config || {}
  return plugin
  /**
   * @type {import('superagent').Plugin}
   */
  function plugin(request) {
    let cfg = Object.assign({}, config)
    let originalEnd = request.end
    /** @type {SuperAgentRequest['end']} */
    request.end = function end(cb) {
      if (cfg.debug) console.log(cfg)
      // only adding prefix/suffix in first request.
      if (!this._redirects) {
        // prefix
        if (cfg.prefix) this.url = cfg.prefix + this.url
        // suffix
        if (cfg.suffix) this.url = this.url + cfg.suffix
      }
      // use plugin before sent
      if (cfg.jsonp && typeof window !== 'undefined') {
        // jsonp will replace end function.
        originalEnd = jsonp(cfg.jsonp)({}).end
      }

      originalEnd.call(this, function callback(
        /** @type {any} */error,
        /** @type {Response} */response) {
        // [response](http://visionmedia.github.io/superagent/#response-properties)
        if (response) {
          // handle 401
          if (response.status === 401 && cfg.unauthorized && typeof window !== 'undefined') {
            if (cfg.unauthorized === true) {
              window.location.reload()
            } else if (typeof cfg.unauthorized === 'string') {
              window.location.href = cfg.unauthorized
            } else if (Array.isArray(cfg.unauthorized)) {
              window.location.href = returnNonNullable(cfg.unauthorized,
                ele => getSubProperty(cfg.debug, response, ele))
            } else if (typeof cfg.unauthorized === 'function') {
              const url = cfg.unauthorized(response)
              if (url) window.location.href = url
            }
          } else if (cfg.report) {
            let reportProperty
            switch (response.statusType) {
              case 2:
                reportProperty = true
                break
              case 4:
                reportProperty = cfg.clientError
                break
              case 5:
                reportProperty = cfg.serverError
                break
              default:
            }
            if (reportProperty) {
              if (reportProperty === true) {
                cfg.report(error, getSubProperty(cfg.debug, response, cfg.messageProperty || 'res.statusMessage'), response)
              } else if (typeof reportProperty === 'string') {
                cfg.report(error, reportProperty, response)
              } else if (Array.isArray(reportProperty)) {
                cfg.report(error, returnNonNullable(reportProperty, ele => getSubProperty(cfg.debug, response, ele)), response)
              } else if (typeof reportProperty === 'function') {
                cfg.report(error, reportProperty(response), response)
              }
            }
          }
        } else {
          if (error && cfg.report) cfg.report(error, error.message, null)
        }

        if (cfg.swallowError) error = null

        if (response && cfg.returnProperty) response = getSubProperty(cfg.debug, response, cfg.returnProperty)

        if (cb) cb(error, response)
      })
    }
    const configure = config.configure || 'configure'
    if (request[configure] !== undefined) {
      throw new Error('configure method name: ' + configure + ', is in conflict.')
    }
    /**
     * @template {Exclude<keyof config, 'configure'>} T
     * @param {T} prop
     * @param {config[T]} value
     */
    request[configure] = function(prop, value) {
      // object
      if (prop === 'jsonp') {
        cfg[prop] = Object.assign({}, cfg[prop], value)
      } else {
        cfg[prop] = value
      }
      return this
    }
  }
}
